pallets/asset-tx-payment/src/lib.rs

```diff
@@ -0,0 +1,256 @@
+pub(crate) type ChargeAssetBalanceOf<T> = <<T as pallet::Config>::OnChargeAssetTransaction as OnChargeAssetTransaction<T>>::Balance;
+pub(crate) type AssetIdOf<T> = <<T as Config>::Fungibles as Inspect<<T as frame_system::Config>::AccountId>>::AssetId;
+pub(crate) type ChargeAssetIdOf<T> = <<T as pallet::Config>::OnChargeAssetTransaction as OnChargeAssetTransaction<T>>::AssetId;
+pub(crate) type LiquidityInfoOf<T> = <<T as pallet_transaction_payment::Config>::OnChargeTransaction as OnChargeTransaction<T>>::LiquidityInfo;
+pub(crate) type ChargeAssetLiquidityOf<T> = <<T as pallet::Config>::OnChargeAssetTransaction as OnChargeAssetTransaction<T>>::LiquidityInfo;
+
+#[derive(Encode, Decode, DefaultNoBound)]
+pub enum InitialPayment<T: Config> {
+	Nothing,
+	Native(LiquidityInfoOf<T>),
+	Asset(CreditOf<T::AccountId, T::Fungibles>),
+}
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	use frame_support::{
+		dispatch::DispatchResultWithPostInfo,
+		pallet_prelude::*,
+		inherent::Vec,
+		traits::{
+			Currency, ReservableCurrency, EnsureOrigin, ExistenceRequirement::KeepAlive,
+		},
+		PalletId,
+	};
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_transaction_payment::Config + pallet_balances::Config {
+		type Fungibles: Balanced<Self::AccountId>;
+		type OnChargeAssetTransaction: OnChargeAssetTransaction<Self>;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {}
+}
+
+/// Require the transactor pay for themselves and maybe include a tip to gain additional priority
+/// in the queue. Allows paying via both `Currency` as well as `fungibles::Balanced`.
+#[derive(Encode, Decode, Clone, Eq, PartialEq)]
+pub struct ChargeAssetTxPayment<T: Config>(#[codec(compact)] BalanceOf<T>, Option<ChargeAssetIdOf<T>>);
+
+impl<T: Config> ChargeAssetTxPayment<T> where
+	T::Call: Dispatchable<Info=DispatchInfo, PostInfo=PostDispatchInfo>,
+	BalanceOf<T>: Send + Sync + FixedPointOperand + IsType<ChargeAssetBalanceOf<T>>,
+	ChargeAssetIdOf<T>: Send + Sync,
+	AssetBalanceOf<T>: Send + Sync + FixedPointOperand,
+	CreditOf<T::AccountId, T::Fungibles>: IsType<ChargeAssetLiquidityOf<T>>,
+{
+	/// utility constructor. Used only in client/factory code.
+	pub fn from(fee: BalanceOf<T>, asset_id: Option<ChargeAssetIdOf<T>>) -> Self {
+		Self(fee, asset_id)
+	}
+
+	fn withdraw_fee(
+		&self,
+		who: &T::AccountId,
+		call: &T::Call,
+		info: &DispatchInfoOf<T::Call>,
+		len: usize,
+	) -> Result<
+		(
+			BalanceOf<T>,
+			InitialPayment<T>,
+		),
+		TransactionValidityError,
+	> {
+		let tip = self.0;
+		let fee = pallet_transaction_payment::Module::<T>::compute_fee(len as u32, info, tip);
```

Comment:
@shawntabrizi We reuse `compute_fee` of `pallet_transaction_payment` here, thus the the dependence on its `Config` trait.

Reply:
Could this be made a standalone function?

Reply:
No, it depends on details of the transaction payment pallet (notably the `FeeMultiplier`).
We could introduce a trait and implement it for the tx payment pallet.

Reply:
I think we def should use a trait here, and not force us to have the pallets be tightly coupled.

We could even duplicate some code here if we should consider this pallet to just be different than the normal transaction pallet?

Reply:
mmh, but they *are* tightly coupled in semantics as well as code. This is just a wrapper around the regular tx payment pallet that extends it. It's not meant to be used with any other pallet.